src/env/test/customization.rs

```diff
@@ -0,0 +1,82 @@
+use std::slice::from_raw_parts;
+
+use crate::api::customization::{Customization, CustomizationImpl};
+use crate::ctap::data_formats::CredentialProtectionPolicy;
+
+pub struct TestCustomization {
+    pub default_min_pin_length: u8,
+    default_min_pin_length_rp_ids_backing_store: Vec<String>,
+    default_min_pin_length_rp_ids: Vec<*const str>,
+    pub default_cred_protect: Option<CredentialProtectionPolicy>,
+    pub max_msg_size: usize,
+    pub max_rp_ids_length: usize,
+}
+
+impl TestCustomization {
+    pub fn set_default_min_pin_length_rp_ids(&mut self, rp_ids: Vec<String>) {
+        self.default_min_pin_length_rp_ids_backing_store = rp_ids;
+        self.default_min_pin_length_rp_ids = self
+            .default_min_pin_length_rp_ids_backing_store
+            .iter()
+            .map(|s| s.as_ref() as *const str)
+            .collect::<Vec<_>>();
+    }
+}
+
+impl Customization for TestCustomization {
+    fn default_cred_protect(&self) -> Option<CredentialProtectionPolicy> {
+        self.default_cred_protect
+    }
+
+    fn default_min_pin_length(&self) -> u8 {
+        self.default_min_pin_length
+    }
+
+    fn default_min_pin_length_rp_ids(&self) -> &[&str] {
```

Comment:
@ia0 @kaczmarczyck is there better way of doing this? The current implementation is quite complicated and I feel it's unnecessary.

Reply:
Yes, I think we should change the `Customization` API:

```rust
// Current PR:
fn default_min_pin_length_rp_ids(&self) -> &[&str];
// Proposal:
fn default_min_pin_length_rp_ids(&self) -> Vec<String>;
```

The cost of creating the vector seems to already be present in `ctap::storage::min_pin_length_rp_ids()`. It's not as clear for `ctap::storage::set_min_pin_length_rp_ids` because if the RP id was already present in the function argument then we recreate it. But it's probably fine since the cost is comparable to the final result.

There are 2 alternatives:
```rust
// This isn't yet supported (but will in the future):
fn default_min_pin_length_rp_ids(&self) -> impl Iterator<Item = &str>;
// This would add rp_ids if they are not already there. It could either be called with an empty Vec to get the list, or with the argument to merge. This is essentially the union of both usages.
fn merge_default_min_pin_length_rp_ids(&self, rp_ids: &mut Vec<String>);
```